File: engines/typescript/src/bundledData.test.ts
/**
 * The catalogs and the quickstart project the engine ships (`data/`) stay byte-identical to their
 * `spec/` and `corpus/` originals, and to the copies the Python and Java engines carry.
 */
import assert from "node:assert/strict";
import { existsSync, readFileSync, readdirSync, statSync } from "node:fs";
import { join, relative } from "node:path";
import { test } from "node:test";

const ENGINE = join(__dirname, "..");
const REPO = join(ENGINE, "..", "..");

function files(root: string): Map<string, Buffer> {
  const found = new Map<string, Buffer>();
  const walk = (dir: string): void => {
    for (const name of readdirSync(dir).sort()) {
      if (name === ".DS_Store" || name === "__pycache__") continue;
      const path = join(dir, name);
      if (statSync(path).isDirectory()) walk(path);
      else found.set(relative(root, path).split("\\").join("/"), readFileSync(path));
    }
  };
  walk(root);
  return found;
}

// (vendored tree, authoritative tree); the overlays tree carries a README the engine does not vendor.
const TREES: [string, string, boolean][] = [
  [join(ENGINE, "data", "catalogs", "base"), join(REPO, "spec", "catalogs", "base"), false],
  [join(ENGINE, "data", "catalogs", "overlays"), join(REPO, "spec", "catalogs", "overlays"), true],
  [join(ENGINE, "data", "corpus", "quickstart"), join(REPO, "corpus", "quickstart"), false],
];

for (const [vendored, authoritative, skipRootReadme] of TREES) {
  test(`the vendored ${relative(ENGINE, vendored)} tree is byte-identical to its original`, () => {
    assert.ok(existsSync(vendored), `${vendored} is missing`);
    const want = files(authoritative);
    if (skipRootReadme) want.delete("README.md");
    const got = files(vendored);
    assert.deepEqual([...got.keys()], [...want.keys()], "the two trees list different files");
    const drifted = [...want.keys()].filter(
      (name) => !got.get(name)?.equals(want.get(name) as Buffer),
    );
    assert.deepEqual(drifted, [], `vendored bytes drifted from ${relative(REPO, authoritative)}`);
  });
}

test("the vendored data is byte-identical to the Python engine's copy", () => {
  const python = join(REPO, "engines", "python", "agentce", "data");
  const pairs: [string, string][] = [
    [join(ENGINE, "data", "catalogs"), join(python, "catalogs")],
    [join(ENGINE, "data", "corpus"), join(python, "corpus")],
  ];
  for (const [ours, theirs] of pairs) {
    const a = files(ours);
    const b = files(theirs);
    assert.deepEqual([...a.keys()], [...b.keys()]);
    for (const [name, bytes] of a) assert.ok(bytes.equals(b.get(name) as Buffer), name);
  }
});

File: engines/java/src/test/java/org/agentce/BundledDataTest.java
package org.agentce;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.stream.Stream;
import org.junit.jupiter.api.Test;

/**
 * The catalogs and the quickstart project the engine ships on its classpath stay byte-identical to
 * their {@code spec/} and {@code corpus/} originals and to the Python engine's copy.
 */
class BundledDataTest {
    private static Map<String, byte[]> files(Path root, boolean skipRootReadme) throws IOException {
        Map<String, byte[]> found = new TreeMap<>();
        try (Stream<Path> walk = Files.walk(root)) {
            for (Path path : (Iterable<Path>) walk.filter(Files::isRegularFile)::iterator) {
                String name = root.relativize(path).toString().replace('\\', '/');
                if (name.equals(".DS_Store") || name.endsWith("/.DS_Store") || (skipRootReadme && name.equals("README.md"))) {
                    continue;
                }
                found.put(name, Files.readAllBytes(path));
            }
        }
        return found;
    }

    private static Path onClasspath(String resource) throws URISyntaxException {
        var url = BundledDataTest.class.getResource(resource);
        assertNotNull(url, resource + " must be on the classpath");
        return Path.of(url.toURI());
    }

    private static void assertSameTree(Path vendored, Path original, boolean skipRootReadme) throws IOException {
        Map<String, byte[]> want = files(original, skipRootReadme);
        Map<String, byte[]> got = files(vendored, false);
        assertEquals(want.keySet(), got.keySet(), vendored + " and " + original + " list different files");
        for (var entry : want.entrySet()) {
            assertArrayEquals(entry.getValue(), got.get(entry.getKey()), "drifted from the original: " + entry.getKey());
        }
    }

    @Test
    void bundledCatalogsAndQuickstartAreByteIdenticalToTheirOriginals() throws IOException, URISyntaxException {
        Path repo = TestPaths.repoRoot();
        assertSameTree(onClasspath("/catalogs/base"), repo.resolve("spec/catalogs/base"), false);
        assertSameTree(onClasspath("/catalogs/overlays"), repo.resolve("spec/catalogs/overlays"), true);
        assertSameTree(onClasspath("/corpus/quickstart"), repo.resolve("corpus/quickstart"), false);
    }

    @Test
    void bundledDataIsByteIdenticalToThePythonEnginesCopy() throws IOException, URISyntaxException {
        Path python = TestPaths.repoRoot().resolve("engines/python/agentce/data");
        for (List<String> pair : List.of(List.of("/catalogs", "catalogs"), List.of("/corpus", "corpus"))) {
            assertSameTree(onClasspath(pair.get(0)), python.resolve(pair.get(1)), false);
        }
    }
}
